Migrations fails during upgrade from 049.8 to 0.50.13

### Describe the bug

Upgrade Metabase from 049.8  to 0.50.13 got the following error.

```
2024-07-17 06:29:16,168 INFO db.liquibase :: Running 55 migrations ...
2024-07-17 06:39:18,093 ERROR liquibase.changelog :: ChangeSet migrations/001_update_migrations.yaml::v50.2024-02-26T22:15:54::noahmoss encountered an exception.
liquibase.exception.DatabaseException: An I/O error occurred while sending to the backend. [Failed SQL: (0) -- Insert table-level view-data permissions when necessary: when a group has
-- table-level data-access permissions, and some of the `no-self-service` rows
-- cannot be automatically migrated to `unrestricted` due to conflicts with
-- other groups
INSERT INTO
  data_permissions (
    group_id,
    perm_type,
    db_id,
    schema_name,
    table_id,
    perm_value
  )
SELECT
  dp.group_id,
  'perms/view-data' AS perm_type,
  dp.db_id,
  dp.schema_name,
  dp.table_id,
  CASE
    WHEN dp.perm_value = 'no-self-service'
    AND EXISTS (
      SELECT
        1
      FROM
        permissions_group_membership pgm
        JOIN (
          SELECT
            group_id,
            db_id,
            CAST(NULL AS INTEGER) AS table_id
          FROM
            connection_impersonations
          UNION
          SELECT
            group_id,
            db_id,
            CAST(NULL AS INTEGER) AS table_id
          FROM
            data_permissions
          WHERE
            perm_value = 'block'
            AND table_id IS NULL
          UNION
          SELECT
            group_id,
            CAST(NULL AS INTEGER) as db_id,
            table_id
          FROM
            sandboxes
        ) AS sp ON pgm.group_id = sp.group_id
      WHERE
        pgm.group_id <> dp.group_id
        AND (
          sp.db_id IS NULL
          OR sp.db_id = dp.db_id
        )
        AND (
          sp.table_id IS NULL
          OR sp.table_id = dp.table_id
        )
    ) THEN 'legacy-no-self-service'
    ELSE 'unrestricted'
  END AS perm_value
FROM
  data_permissions dp
WHERE
  dp.perm_type = 'perms/data-access'
  AND dp.table_id IS NOT NULL]
        at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:470)
        at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:77)
        at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:179)
        at liquibase.executor.AbstractExecutor.execute(AbstractExecutor.java:141)
        at liquibase.database.AbstractJdbcDatabase.executeStatements(AbstractJdbcDatabase.java:1285)
        at liquibase.changelog.ChangeSet.execute(ChangeSet.java:755)
        at liquibase.changelog.visitor.UpdateVisitor.executeAcceptedChange(UpdateVisitor.java:119)
        at liquibase.changelog.visitor.UpdateVisitor.visit(UpdateVisitor.java:68)
        at liquibase.changelog.ChangeLogIterator$2.lambda$run$0(ChangeLogIterator.java:133)
        at liquibase.Scope.lambda$child$0(Scope.java:186)
        at liquibase.Scope.child(Scope.java:195)
        at liquibase.Scope.child(Scope.java:185)
        at liquibase.Scope.child(Scope.java:164)
        at liquibase.changelog.ChangeLogIterator$2.run(ChangeLogIterator.java:122)
        at liquibase.Scope.lambda$child$0(Scope.java:186)
        at liquibase.Scope.child(Scope.java:195)
        at liquibase.Scope.child(Scope.java:185)
        at liquibase.Scope.child(Scope.java:164)
        at liquibase.Scope.child(Scope.java:252)
        at liquibase.Scope.child(Scope.java:256)
        at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:89)
        at liquibase.command.core.AbstractUpdateCommandStep.lambda$run$0(AbstractUpdateCommandStep.java:110)
        at liquibase.Scope.lambda$child$0(Scope.java:186)
        at liquibase.Scope.child(Scope.java:195)
        at liquibase.Scope.child(Scope.java:185)
        at liquibase.Scope.child(Scope.java:164)
        at liquibase.command.core.AbstractUpdateCommandStep.run(AbstractUpdateCommandStep.java:108)
        at liquibase.command.core.UpdateCommandStep.run(UpdateCommandStep.java:105)
        at liquibase.command.CommandScope.execute(CommandScope.java:217)
        at liquibase.Liquibase.lambda$update$0(Liquibase.java:245)
        at liquibase.Scope.lambda$child$0(Scope.java:186)
        at liquibase.Scope.child(Scope.java:195)
        at liquibase.Scope.child(Scope.java:185)
        at liquibase.Scope.child(Scope.java:164)
        at liquibase.Liquibase.runInScope(Liquibase.java:1419)
        at liquibase.Liquibase.update(Liquibase.java:234)
        at liquibase.Liquibase.update(Liquibase.java:212)
        at liquibase.Liquibase.update(Liquibase.java:194)
        at metabase.db.liquibase$migrate_up_if_needed_BANG_$fn__44519.invoke(liquibase.clj:360)
        at metabase.db.liquibase$run_in_scope_locked$reify__44515.run(liquibase.clj:325)
        at liquibase.Scope.lambda$child$0(Scope.java:186)
        at liquibase.Scope.child(Scope.java:195)
        at liquibase.Scope.child(Scope.java:185)
        at liquibase.Scope.child(Scope.java:164)
        at metabase.db.liquibase$run_in_scope_locked.invokeStatic(liquibase.clj:318)
        at metabase.db.liquibase$run_in_scope_locked.invoke(liquibase.clj:301)
        at metabase.db.liquibase$migrate_up_if_needed_BANG_.invokeStatic(liquibase.clj:349)
        at metabase.db.liquibase$migrate_up_if_needed_BANG_.invoke(liquibase.clj:342)
        at metabase.db.setup$migrate_BANG_$fn__53455.invoke(setup.clj:84)
        at metabase.db.liquibase$do_with_liquibase$f_STAR___44456.invoke(liquibase.clj:140)
        at metabase.db.liquibase$do_with_liquibase.invokeStatic(liquibase.clj:143)
        at metabase.db.liquibase$do_with_liquibase.invoke(liquibase.clj:131)
        at metabase.db.setup$migrate_BANG_.invokeStatic(setup.clj:73)
        at metabase.db.setup$migrate_BANG_.doInvoke(setup.clj:55)
        at clojure.lang.RestFn.invoke(RestFn.java:425)
        at metabase.db.setup$run_schema_migrations_BANG_.invokeStatic(setup.clj:149)
        at metabase.db.setup$run_schema_migrations_BANG_.invoke(setup.clj:144)
        at metabase.db.setup$setup_db_BANG_$fn__53483$fn__53484.invoke(setup.clj:167)
        at metabase.util.jvm$do_with_us_locale.invokeStatic(jvm.clj:239)
        at metabase.util.jvm$do_with_us_locale.invoke(jvm.clj:225)
        at metabase.db.setup$setup_db_BANG_$fn__53483.invoke(setup.clj:161)
        at metabase.db.setup$setup_db_BANG_.invokeStatic(setup.clj:160)
        at metabase.db.setup$setup_db_BANG_.invoke(setup.clj:153)
        at metabase.db$setup_db_BANG_$fn__53508.invoke(db.clj:86)
        at metabase.db$setup_db_BANG_.invokeStatic(db.clj:81)
        at metabase.db$setup_db_BANG_.doInvoke(db.clj:68)
        at clojure.lang.RestFn.invoke(RestFn.java:421)
        at metabase.core$init_BANG__STAR_.invokeStatic(core.clj:117)
        at metabase.core$init_BANG__STAR_.invoke(core.clj:98)
        at metabase.core$init_BANG_.invokeStatic(core.clj:170)
        at metabase.core$init_BANG_.invoke(core.clj:165)
        at metabase.core$start_normally.invokeStatic(core.clj:182)
        at metabase.core$start_normally.invoke(core.clj:176)
        at metabase.core$entrypoint.invokeStatic(core.clj:215)
        at metabase.core$entrypoint.doInvoke(core.clj:209)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at clojure.lang.Var.applyTo(Var.java:705)
        at clojure.core$apply.invokeStatic(core.clj:667)
        at clojure.core$apply.invoke(core.clj:662)
        at metabase.bootstrap$_main.invokeStatic(bootstrap.clj:31)
        at metabase.bootstrap$_main.doInvoke(bootstrap.clj:28)
        at clojure.lang.RestFn.invoke(RestFn.java:397)
        at clojure.lang.AFn.applyToHelper(AFn.java:152)
        at clojure.lang.RestFn.applyTo(RestFn.java:132)
        at metabase.bootstrap.main(Unknown Source)
Caused by: org.postgresql.util.PSQLException: An I/O error occurred while sending to the backend.
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:398)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:502)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:419)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:341)
        at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:326)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:302)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:297)
        at com.mchange.v2.c3p0.impl.NewProxyStatement.execute(NewProxyStatement.java:75)
        at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:464)
        ... 86 more
Caused by: java.io.EOFException
        at org.postgresql.core.PGStream.receiveChar(PGStream.java:469)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2166)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:371)
        ... 94 more

UPDATE SUMMARY
Run:                         55
Previously run:             295
Filtered out:                50
-------------------------------
Total change sets:          400


FILTERED CHANGE SETS SUMMARY
DBMS mismatch:               50

```

The `ERROR liquibase.changelog` error happened after 10 min migration stared. Probably timeout?

### To Reproduce

Change image in docker run command.

### Expected behavior

Migrations successfully finished

### Logs

_No response_

### Information about your Metabase installation

```JSON
- Metabase in docker on a VM
- Database Postgres 16.1 AWS RDS
```


### Severity

-

### Additional context

_No response_

--- a/resources/migrations/001_update_migrations.yaml
+++ b/resources/migrations/001_update_migrations.yaml
@@ -7021,6 +7021,52 @@ databaseChangeLog:
               DROP VIEW IF EXISTS v_content;
 
   - changeSet:
+      id: v50.2024-02-26T22:15:52
+      author: noahmoss
+      comment: Add index on data_permissions(group_id, db_id, perm_value)
+      preConditions:
+        - onFail: MARK_RAN
+        - not:
+            - indexExists:
+                tableName: data_permissions
+                indexName: idx_data_permissions_group_id_db_id_perm_value
+      changes:
+        - createIndex:
+            tableName: data_permissions
+            indexName: idx_data_permissions_group_id_db_id_perm_value
+            columns:
+              - column:
+                  name: group_id
+              - column:
+                  name: db_id
+              - column:
+                  name: perm_value
+
+  - changeSet:
+      id: v50.2024-02-26T22:15:53
+      author: noahmoss
+      comment: Add index on data_permissions(group_id, db_id, table_id, perm_value)
+      preConditions:
+        - onFail: MARK_RAN
+        - not:
+            - indexExists:
+                tableName: data_permissions
+                indexName: idx_data_permissions_group_id_db_id_table_id_perm_value
+      changes:
+        - createIndex:
+            tableName: data_permissions
+            indexName: idx_data_permissions_group_id_db_id_table_id_perm_value
+            columns:
+              - column:
+                  name: group_id
+              - column:
+                  name: db_id
+              - column:
+                  name: table_id
+              - column:
+                  name: perm_value
+
+  - changeSet:
       id: v50.2024-02-26T22:15:54
       author: noahmoss
       comment: New `view-data` permission

--- a/resources/migrations/001_update_migrations.yaml
+++ b/resources/migrations/001_update_migrations.yaml
@@ -7042,6 +7042,7 @@ databaseChangeLog:
       id: v50.2024-02-26T22:15:55
       author: noahmoss
       comment: New `create_queries` permission
+      validCheckSum: 9:7c2bc517b6def4e3278394b0399c3d4b
       changes:
         - sqlFile:
             path: permissions/create_queries.sql

--- a/resources/migrations/001_update_migrations.yaml
+++ b/resources/migrations/001_update_migrations.yaml
@@ -7070,6 +7070,7 @@ databaseChangeLog:
       id: v50.2024-02-26T22:15:54
       author: noahmoss
       comment: New `view-data` permission
+      validCheckSum: 9:68d8f2fdf00395a9e614b0db8a2ec7bb
       changes:
         - sqlFile:
             dbms: postgresql,h2

--- a/resources/migrations/permissions/create_queries.sql
+++ b/resources/migrations/permissions/create_queries.sql
@@ -41,37 +41,26 @@ WHERE pg.name != 'Administrators'
        AND dp.table_id IS NULL
        AND dp.perm_type = 'perms/create-queries' );
 
-
+-- Insert table-level rows for all tables that have table-level data access permissions,
+-- and don't already have any create-queries permissions stored at the DB- or table-level.
 INSERT INTO data_permissions (group_id, perm_type, db_id, schema_name, table_id, perm_value)
-SELECT pg.id AS group_id,
-       'perms/create-queries' AS perm_type,
-       mt.db_id AS db_id,
-       mt.schema AS schema_name,
-       mt.id AS table_id,
-       CASE
-           WHEN EXISTS
-                  (SELECT 1
-                   FROM data_permissions dp
-                   WHERE dp.group_id = pg.id
-                     AND dp.table_id = mt.id
-                     AND dp.perm_type = 'perms/data-access'
-                     AND dp.perm_value = 'unrestricted' ) THEN 'query-builder'
-           ELSE 'no'
-       END AS perm_value
-FROM permissions_group pg
-CROSS JOIN metabase_table mt
-WHERE pg.name != 'Administrators'
-  -- Insert table-level rows for all tables that have table-level data access permissions,
-  -- and don't already have any create-queries permissions stored at the DB- or table-level.
-  AND EXISTS
-    (SELECT 1
-     FROM data_permissions dp
-     WHERE dp.group_id = pg.id
-       AND dp.table_id = mt.id
-       AND dp.perm_type = 'perms/data-access' )
+SELECT
+  dp.group_id,
+  'perms/create-queries' AS perm_type,
+  dp.db_id,
+  dp.schema_name,
+  dp.table_id,
+  CASE
+    WHEN dp.perm_value = 'unrestricted' THEN 'query-builder'
+    ELSE 'no'
+  END AS perm_value
+FROM data_permissions dp
+  WHERE
+    dp.perm_type = 'perms/data-access'
+    AND dp.table_id IS NOT NULL
   AND NOT EXISTS
     (SELECT 1
-     FROM data_permissions dp
-     WHERE dp.group_id = pg.id
-       AND dp.db_id = mt.db_id
-       AND dp.perm_type = 'perms/create-queries' );
+     FROM data_permissions dp2
+     WHERE dp2.group_id = dp.group_id
+       AND dp2.db_id = dp.db_id
+       AND dp2.perm_type = 'perms/create-queries' );

--- a/resources/migrations/permissions/create_queries.sql
+++ b/resources/migrations/permissions/create_queries.sql
@@ -55,12 +55,12 @@ SELECT
     ELSE 'no'
   END AS perm_value
 FROM data_permissions dp
-  WHERE
-    dp.perm_type = 'perms/data-access'
-    AND dp.table_id IS NOT NULL
-  AND NOT EXISTS
-    (SELECT 1
-     FROM data_permissions dp2
-     WHERE dp2.group_id = dp.group_id
-       AND dp2.db_id = dp.db_id
-       AND dp2.perm_type = 'perms/create-queries' );
+WHERE
+  dp.perm_type = 'perms/data-access'
+  AND dp.table_id IS NOT NULL
+AND NOT EXISTS
+  (SELECT 1
+   FROM data_permissions dp2
+   WHERE dp2.group_id = dp.group_id
+     AND dp2.db_id = dp.db_id
+     AND dp2.perm_type = 'perms/create-queries' );

--- a/resources/migrations/permissions/mysql_view_data.sql
+++ b/resources/migrations/permissions/mysql_view_data.sql
@@ -17,56 +17,86 @@ SELECT
   dp.db_id,
   dp.schema_name,
   dp.table_id,
-  CASE
-    WHEN dp.perm_value = 'no-self-service'
-    AND EXISTS (
-      SELECT
-        1
-      FROM
-        permissions_group_membership pgm
-        JOIN (
-          SELECT
-            group_id,
-            db_id,
-            CAST(NULL AS UNSIGNED) AS table_id
-          FROM
-            connection_impersonations
-          UNION
-          SELECT
-            group_id,
-            db_id,
-            CAST(NULL AS UNSIGNED) AS table_id
-          FROM
-            data_permissions
-          WHERE
-            perm_value = 'block'
-            AND table_id IS NULL
-          UNION
-          SELECT
-            group_id,
-            CAST(NULL AS UNSIGNED) as db_id,
-            table_id
-          FROM
-            sandboxes
-        ) AS sp ON pgm.group_id = sp.group_id
-      WHERE
-        pgm.group_id <> dp.group_id
-        AND (
-          sp.db_id IS NULL
-          OR sp.db_id = dp.db_id
-        )
-        AND (
-          sp.table_id IS NULL
-          OR sp.table_id = dp.table_id
-        )
-    ) THEN 'legacy-no-self-service'
-    ELSE 'unrestricted'
-  END AS perm_value
+  'legacy-no-self-service' AS perm_value
 FROM
   data_permissions dp
 WHERE
-  dp.perm_type = 'perms/data-access'
-  AND dp.table_id IS NOT NULL;
+  dp.table_id IS NOT NULL
+  AND dp.perm_type = 'perms/data-access'
+  AND dp.perm_value = 'no-self-service'
+  AND EXISTS (
+    SELECT
+      1
+    FROM
+      permissions_group_membership pgm
+      JOIN (
+        SELECT
+          group_id,
+          db_id,
+          CAST(NULL AS UNSIGNED) AS table_id
+        FROM
+          connection_impersonations
+        UNION
+        SELECT
+          group_id,
+          db_id,
+          CAST(NULL AS UNSIGNED) AS table_id
+        FROM
+          data_permissions
+        WHERE
+          perm_value = 'block'
+          AND table_id IS NULL
+        UNION
+        SELECT
+          group_id,
+          CAST(NULL AS UNSIGNED) as db_id,
+          table_id
+        FROM
+          sandboxes
+      ) AS sp ON pgm.group_id = sp.group_id
+    WHERE
+      pgm.group_id <> dp.group_id
+      AND (
+        sp.db_id IS NULL
+        OR sp.db_id = dp.db_id
+      )
+      AND (
+        sp.table_id IS NULL
+        OR sp.table_id = dp.table_id
+      )
+  );
+
+INSERT INTO
+  data_permissions (
+    group_id,
+    perm_type,
+    db_id,
+    schema_name,
+    table_id,
+    perm_value
+  )
+SELECT
+  dp.group_id,
+  'perms/view-data' AS perm_type,
+  dp.db_id,
+  dp.schema_name,
+  dp.table_id,
+  'unrestricted' AS perm_value
+FROM
+  data_permissions dp
+WHERE
+  dp.table_id IS NOT NULL
+  AND dp.perm_type = 'perms/data-access'
+  AND NOT EXISTS (
+    SELECT
+      1
+    FROM
+    data_permissions dp1
+    WHERE
+     dp1.group_id = dp.group_id
+     AND dp1.table_id = dp.table_id
+     AND dp1.perm_value = 'legacy-no-self-service'
+  );
 
 
 -- If all tables in a DB have the same view-data permission, insert a DB-level view-data permission instead

--- a/resources/migrations/permissions/rollback/view_data.sql
+++ b/resources/migrations/permissions/rollback/view_data.sql
@@ -118,3 +118,11 @@ WHERE pg.name != 'Administrators'
      WHERE dp.group_id = pg.id
        AND dp.table_id = mt.id
        AND dp.perm_type = 'perms/create-queries');
+
+DELETE
+FROM data_permissions
+WHERE perm_type = 'perms/view-data';
+
+DELETE
+FROM data_permissions
+WHERE perm_type = 'perms/create-queries';

--- a/resources/migrations/permissions/view_data.sql
+++ b/resources/migrations/permissions/view_data.sql
@@ -17,57 +17,86 @@ SELECT
   dp.db_id,
   dp.schema_name,
   dp.table_id,
-  CASE
-    WHEN dp.perm_value = 'no-self-service'
-    AND EXISTS (
-      SELECT
-        1
-      FROM
-        permissions_group_membership pgm
-        JOIN (
-          SELECT
-            group_id,
-            db_id,
-            CAST(NULL AS INTEGER) AS table_id
-          FROM
-            connection_impersonations
-          UNION
-          SELECT
-            group_id,
-            db_id,
-            CAST(NULL AS INTEGER) AS table_id
-          FROM
-            data_permissions
-          WHERE
-            perm_value = 'block'
-            AND table_id IS NULL
-          UNION
-          SELECT
-            group_id,
-            CAST(NULL AS INTEGER) as db_id,
-            table_id
-          FROM
-            sandboxes
-        ) AS sp ON pgm.group_id = sp.group_id
-      WHERE
-        pgm.group_id <> dp.group_id
-        AND (
-          sp.db_id IS NULL
-          OR sp.db_id = dp.db_id
-        )
-        AND (
-          sp.table_id IS NULL
-          OR sp.table_id = dp.table_id
-        )
-    ) THEN 'legacy-no-self-service'
-    ELSE 'unrestricted'
-  END AS perm_value
+  'legacy-no-self-service' AS perm_value
 FROM
   data_permissions dp
 WHERE
-  dp.table_id IS NOT NULL;
-  dp.perm_type = 'perms/data-access'
+  dp.table_id IS NOT NULL
+  AND dp.perm_type = 'perms/data-access'
+  AND dp.perm_value = 'no-self-service'
+  AND EXISTS (
+    SELECT
+      1
+    FROM
+      permissions_group_membership pgm
+      JOIN (
+        SELECT
+          group_id,
+          db_id,
+          CAST(NULL AS INTEGER) AS table_id
+        FROM
+          connection_impersonations
+        UNION
+        SELECT
+          group_id,
+          db_id,
+          CAST(NULL AS INTEGER) AS table_id
+        FROM
+          data_permissions
+        WHERE
+          perm_value = 'block'
+          AND table_id IS NULL
+        UNION
+        SELECT
+          group_id,
+          CAST(NULL AS INTEGER) as db_id,
+          table_id
+        FROM
+          sandboxes
+      ) AS sp ON pgm.group_id = sp.group_id
+    WHERE
+      pgm.group_id <> dp.group_id
+      AND (
+        sp.db_id IS NULL
+        OR sp.db_id = dp.db_id
+      )
+      AND (
+        sp.table_id IS NULL
+        OR sp.table_id = dp.table_id
+      )
+  );
 
+INSERT INTO
+  data_permissions (
+    group_id,
+    perm_type,
+    db_id,
+    schema_name,
+    table_id,
+    perm_value
+  )
+SELECT
+  dp.group_id,
+  'perms/view-data' AS perm_type,
+  dp.db_id,
+  dp.schema_name,
+  dp.table_id,
+  'unrestricted' AS perm_value
+FROM
+  data_permissions dp
+WHERE
+  dp.table_id IS NOT NULL
+  AND dp.perm_type = 'perms/data-access'
+  AND NOT EXISTS (
+    SELECT
+      1
+    FROM
+    data_permissions dp1
+    WHERE
+     dp1.group_id = dp.group_id
+     AND dp1.table_id = dp.table_id
+     AND dp1.perm_value = 'legacy-no-self-service'
+  );
 
 -- If all tables in a DB have the same view-data permission, insert a DB-level view-data permission instead
 INSERT INTO

--- a/resources/migrations/permissions/view_data.sql
+++ b/resources/migrations/permissions/view_data.sql
@@ -65,8 +65,8 @@ SELECT
 FROM
   data_permissions dp
 WHERE
+  dp.table_id IS NOT NULL;
   dp.perm_type = 'perms/data-access'
-  AND dp.table_id IS NOT NULL;
 
 
 -- If all tables in a DB have the same view-data permission, insert a DB-level view-data permission instead
@@ -88,8 +88,8 @@ WITH
     FROM
       data_permissions
     WHERE
-      perm_type = 'perms/view-data'
-      AND table_id IS NOT NULL
+      table_id IS NOT NULL
+      AND perm_type = 'perms/view-data'
     GROUP BY
       group_id,
       db_id

